### DISCUSSION TITLE
DAG Policy not working with default args.

### DISCUSSION THREAD

**Discussion by mariotaddeucci** [apache/airflow#24562] [Q&A] [answered]
### Apache Airflow version

2.3.2 (latest released)

### What happened

When changing the `default_args` on DAG using cluster policy, nothing happens.
I think `apply_defaults` on BaseOperator is being called before the dag_policy.

### What you think should happen instead

The cluster policies overwrite the defined values, the same should happen with default_args.

### How to reproduce

Just change a value on default args using cluster policies.

```
def dag_policy(dag: DAG):
    dag.default_args["owner"]="not_working"
    dag.tags = ["working"]
```

### Operating System

Official docker image

### Versions of Apache Airflow Providers

No providers added.

### Deployment

Other

### Deployment details

It happen on airflow standalone, no specific setup to reproduce that.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by mik-laj** [comment]
Yes. Default args are resolved when the object is created in the operator's constructor, and the DAG Policy is executed for the entire loaded DAG, so this will never work and it is expected by design. Would you like to add information about this limitation to the documentation?

**Comment by mariotaddeucci** [comment]
@mik-laj sure, I'll do it.
Is there a reason to not run the policy before? If it makes sense to run before, I can work on it, I'll just need a support for it.

**Comment by mik-laj** [comment, marked as answer]
> Is there a reason to not run the policy before?

Yes. We don't want to introduce an indirect representation of operators that will contain some arguments, but not a fully initialized object. This means that the operator must always be created in one step and it must be an "atomic" operation. To do it, we use [`apply_default`](https://github.com/apache/airflow/blob/2.1.0/airflow/models/baseoperator.py#L140-L188) operators that mutate constructor parameters. 

On the other hand, we want the DAG policy to be executed only once for one DAG.

**Comment by mariotaddeucci** [comment]
Great, I got it. I'll add a doc about it.

**Comment by mariotaddeucci** [comment]
The last question (It's not a promise 😂)
There are many default values on airflow.cfg, instead of adding another parameters, what do you think about add all this parameters into a json cfg and load it, then will be easier to add new default args
I know cluster policies can solve it, but already I'm getting some issues with cluster policies and task mapping, and defining the values on default args could solve it.

**Reply by potiuk** [reply]
The default_args are python code that needs to be executed. It's not a json dict. this is something that can (and often does) contain python code to execute to provide the right values for defaults. And the value provided by those should not be serialized or stored - it can even contains things like local host name or IP address that you might want to use in your task. 

Simply speaking - default_args are dynamically derived a the moment the task object is created - in the context of the worker where the task is executed and at the exact time when the task is instantiated.

It can even happen that the task - when retried - will have DIFFERENT default_args (for example because it will be retried on a diferent worker).

**Reply by mik-laj** [reply]
It is possible to define several defaults with the `airflow.cfg` file, but this is a very narrow and limited subset of parameters that affect how the scheduler behaves.  I don't think it makes sense to add another way to set defaults this time with a JSON value. JSON supports a limited set of data types, but parameters can be any Python objects.
